### ISSUE TITLE
Impossible to send uploaded encrypted file < 10MB because of Md5ChecksumInvalidError

### ISSUE THREAD

**Issue by NotStatilko** [LonamiWebs/Telethon#3023]
Hello. I can't send encrypted and uploaded (with `upload_file`) by Telethon file because Telethon computes md5 hash of file **before** the encryption, so it seems that after uploading it Telethon computes hash of **encrypted** file and fails with check.

There is **no** such problem with files bigger than 10MB so maybe a bug?

https://github.com/LonamiWebs/Telethon/blob/4b16183d2bbe80cbf4dabdb266a8015c5bf975cc/telethon/client/uploads.py#L609-L617

Test code (raises `Md5ChecksumInvalidError`):
```python3
from telethon import TelegramClient
from asyncio import run as asyncio_run
from os import urandom

API_ID = 'api_id'
API_HASH = 'api_hash'

async def main():
    client = TelegramClient('session', API_ID, API_HASH)
    await client.start()

    ufile = await client.upload_file(
        open('m.mp3','rb'), # m.mp3 is < 10MB 
        key=urandom(32), 
        iv=urandom(32), 
        part_size_kb=512
    ) 
    await client.send_file('me',file=ufile)


asyncio_run(main())
```

**Comment by Lonami** [comment]
Telegram does not use the MD5 for files larger than 10MB. Does it work if you move the `if not is_big` block below the `if key and iv` block you linked? If it does fix it, would you mind sending a PR?

**Comment by EvilMegaDroid** [comment]
This isssue is not fixed, I can still hit it with the latest Telethon commits (pip install from master).

**Comment by Lonami** [comment]
Related: #3305.

**Comment by EvilMegaDroid** [comment]
While checking an account running Telethon I found out this error (which brought me here)

```
Traceback (most recent call last):
  File "/app/app/bot.py", line 3015, in _send_message
    sent_msg = await self.send_message(
  File "/app/app/bot.py", line 2840, in send_message
    return await super().send_message(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/telethon/client/messages.py", line 782, in send_message
    return await self.send_file(
  File "/usr/local/lib/python3.8/site-packages/telethon/client/uploads.py", line 394, in send_file
    return self._get_response_message(request, await self(request), entity)
  File "/app/app/bot.py", line 652, in __call__
    return await super().__call__(request, ordered)
  File "/usr/local/lib/python3.8/site-packages/telethon/client/users.py", line 30, in __call__
    return await self._call(self._sender, request, ordered=ordered)
  File "/usr/local/lib/python3.8/site-packages/telethon/client/users.py", line 79, in _call
    result = await future
telethon.errors.rpcerrorlist.Md5ChecksumInvalidError: The MD5 check-sums do not match (caused by SendMediaRequest)
```

**Comment by NotStatilko** [comment]
> Related: #3035.

No, it's only about random padding. This (#3023) issue is just about small bug with hashing.

**Comment by NotStatilko** [comment]
@EvilMegaDroid can you show a part of your code with file uploading?

**Comment by EvilMegaDroid** [comment]
@NotStatilko Fixed this issue, the fault was in my own code.
I found a tweak online which improves Telethon download and upload performance (a gist called fast_telethon) and decided to try and it since it really improved speed decided to use it.

It seems that the upload method on that was faulty and it was causing this issue. Since the trackback did not have it I though that was not the issue initially.

You can close this issue

**Comment by NotStatilko** [comment]
> ...which improves Telethon download and upload performance

By default Telethon uses `pyaes` for encryption, do you tried to install `cryptg` additionally and test your code again?

`python3 -m pip install cryptg`

**Comment by EvilMegaDroid** [comment]
Hey, yes I have cryptg install.
The gist I talked about I think makes use of parallel downloading to download the files faster.

One thing I noticed is that using it with the default workers the gist had it seems that telegram would disconnect the user. Then Telethon would raise the #1634 issue, maybe using that gist to track that issue might help.

Here's the gist: https://gist.github.com/painor/7e74de80ae0c819d3e9abcf9989a8dd6

Using 15-20 workers always raises issue #1634 in a few seconds
